From 639e475a4e36bf339b63741ccda9c647ecfa83aa Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 9 Jun 2023 15:12:31 +0800
Subject: [PATCH 0584/1533] fsync: Type-check HANDLE in fsync_reset_event().

Oddworld: Stranger's Wrath HD (15750) calls ResetEvent() on a file handle and then waits for the
handle with an infinite timeout. Without esync/fsync, NtResetEvent() should return
STATUS_OBJECT_TYPE_MISMATCH because the handle is not an event handle. With esync/fsync, the file
handle is set to non-signaled successfully and causes the game to hang at start-up. The same check
should also apply to fsync_pulse_event(), esync_reset_event(), and esync_pulse_event().

CW-Bug-Id: #22326
---
 dlls/ntdll/unix/fsync.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/ntdll/unix/fsync.c b/dlls/ntdll/unix/fsync.c
index 8f20f7cab4a..3dfa7c4e035 100644
--- a/dlls/ntdll/unix/fsync.c
+++ b/dlls/ntdll/unix/fsync.c
@@ -714,6 +714,12 @@ NTSTATUS fsync_reset_event( HANDLE handle, LONG *prev )
     if ((ret = get_object( handle, &obj ))) return ret;
     event = obj.shm;
 
+    if (obj.type != FSYNC_MANUAL_EVENT && obj.type != FSYNC_AUTO_EVENT)
+    {
+        put_object( &obj );
+        return STATUS_OBJECT_TYPE_MISMATCH;
+    }
+
     current = __atomic_exchange_n( &event->signaled, 0, __ATOMIC_SEQ_CST );
 
     if (prev) *prev = current;
-- 
2.45.0

